File: requests.py
import shutil
import http.client
import urllib.parse
import ssl
import logging

logging.basicConfig(level=logging.INFO, format="%(message)s")
logger = logging.getLogger(__name__)

def request(host, url, method="GET", headers={}, save_path=None):
    """
    General function for performing HTTP/HTTPS requests with manual redirect handling (3xx).
    Logic copied from DockerPuller._make_request.
    """
    MAX_REDIRECTS = 5
    redirect_count = 0
    current_host = host
    current_url = url

    while redirect_count < MAX_REDIRECTS:
        conn = None
        try:
            parsed_url = urllib.parse.urlparse(f"https://{current_host}{current_url}")
            current_host = parsed_url.netloc
            current_path = parsed_url.path + (
                "?" + parsed_url.query if parsed_url.query else ""
            )

            context = ssl.create_default_context()
            conn = http.client.HTTPSConnection(current_host, context=context)

            # KEY LOGIC FROM YOUR CLASS: Removing the Authorization header after the first redirect
            req_headers = headers.copy()
            if redirect_count > 0 and "Authorization" in req_headers:
                # After redirecting to the storage server (blobs), authorization is often built into the link
                del req_headers["Authorization"]

            conn.request(method, current_path, headers=req_headers)
            response = conn.getresponse()

            # Handling redirects (Status 3xx)
            if response.status in (301, 302, 307, 308):
                new_location = response.getheader("Location")
                if not new_location:
                    raise Exception("Redirect without Location header.")

                new_parsed_url = urllib.parse.urlparse(new_location)
                current_host = new_parsed_url.netloc
                current_url = new_parsed_url.path + (
                    "?" + new_parsed_url.query if new_parsed_url.query else ""
                )

                redirect_count += 1
                conn.close()
                continue

            if response.status == 200:
                if save_path:
                    with open(save_path, "wb") as f:
                        shutil.copyfileobj(response, f)
                    return "File saved", 200
                else:
                    data = response.read().decode("utf-8")
                    return data, 200
            else:
                error_data = response.read().decode("utf-8", errors="ignore")
                logger.error(
                    f"HTTP Status {response.status} for {current_host}{current_path}"
                )
                return error_data, response.status

        except Exception as e:
            logger.error(f"Error during request to {current_host}{current_path}: {e}")
            return None, None
        finally:
            if conn:
                conn.close()

    if redirect_count == MAX_REDIRECTS:
        logger.error("Maximum redirect limit reached.")
        return None, None

    return None, None

File: pkg/api/requests.go
package api

import (
	"crypto/tls"
	"fmt"
	"io"
	"net/http"
	"os"
)

func Request(host, path, method string, headers map[string]string, savePath string) (string, int, error) {
	maxRedirects := 5
	redirectCount := 0
	currentURL := fmt.Sprintf("https://%s%s", host, path)

	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
		},
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	for redirectCount < maxRedirects {
		req, err := http.NewRequest(method, currentURL, nil)
		if err != nil {
			return "", 0, err
		}

		req.Header.Set("User-Agent", "container-manager/1.0")
		for k, v := range headers {
			if redirectCount > 0 && k == "Authorization" {
				continue
			}
			req.Header.Set(k, v)
		}

		resp, err := client.Do(req)
		if err != nil {
			return "", 0, err
		}
		defer resp.Body.Close()

		if resp.StatusCode >= 300 && resp.StatusCode < 400 {
			location := resp.Header.Get("Location")
			if location == "" {
				return "", resp.StatusCode, fmt.Errorf("redirect without location")
			}
			currentURL = location
			redirectCount++
			continue
		}

		if resp.StatusCode == 200 {
			if savePath != "" {
				file, err := os.Create(savePath)
				if err != nil {
					return "", resp.StatusCode, err
				}
				defer file.Close()
				_, err = io.Copy(file, resp.Body)
				return "File saved", resp.StatusCode, err
			}

			body, err := io.ReadAll(resp.Body)
			return string(body), resp.StatusCode, err
		}

		body, _ := io.ReadAll(resp.Body)
		return string(body), resp.StatusCode, nil
	}

	return "", 0, fmt.Errorf("max redirects reached")
}
